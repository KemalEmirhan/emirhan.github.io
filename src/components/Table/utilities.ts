interface TableData {
  [key: string]: string;
}

export const isLinkField = (
  header: string,
  linkField: string | undefined
): boolean =>
  Boolean(linkField && header.toLowerCase() === linkField.toLowerCase());

export const getValue = (
  row: TableData,
  header: string,
  isLeft: boolean
): string => {
  const key = header.toLowerCase();
  return (
    row[key] || (isLeft ? row[Object.keys(row)[0]] : row[Object.keys(row)[1]])
  );
};
