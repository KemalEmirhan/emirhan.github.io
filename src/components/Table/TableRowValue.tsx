'use client';

import React from 'react';
import Link from 'next/link';
import { ArrowUpRight } from 'lucide-react';
import { isLinkField } from './utilities';

interface TableRowValueProps {
  value: string;
  header: string;
  linkField?: string;
  direction: 'left' | 'right';
}

const TableRowValue = ({
  value,
  header,
  linkField,
  direction,
}: TableRowValueProps): React.ReactElement => {
  if (isLinkField(header, linkField)) {
    return (
      <Link
        href={value}
        target='_blank'
        rel='noopener noreferrer'
        className='text-gray-500 hover:text-gray-800 underline flex items-center text-base'
      >
        {value}
        {direction === 'right' && <ArrowUpRight className='w-4 h-4' />}
      </Link>
    );
  }

  return <span className='text-base uppercase'>{value}</span>;
};

export default TableRowValue;
