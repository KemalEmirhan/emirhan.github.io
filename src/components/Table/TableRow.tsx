import React from 'react';
import { getValue } from './utilities';
import TableRowValue from './TableRowValue';

interface TableData {
  [key: string]: string;
}

interface TableRowProps {
  row: TableData;
  leftHeader: string;
  rightHeader: string;
  linkField?: string;
}

const TableRow = ({
  row,
  leftHeader,
  rightHeader,
  linkField,
}: TableRowProps): React.ReactElement => {
  const leftValue = getValue(row, leftHeader, true);
  const rightValue = getValue(row, rightHeader, false);

  return (
    <div className='px-2 md:px-4 lg:px-0'>
      <div className='flex items-baseline'>
        <div className='text-left pr-2'>
          <TableRowValue
            value={leftValue}
            header={leftHeader}
            linkField={linkField}
            direction='left'
          />
        </div>
        <div className='flex-1 relative overflow-hidden h-[2px]'>
          <div className='absolute inset-0 flex items-center dot-leaders' />
        </div>
        <div className='text-right pl-2'>
          <TableRowValue
            value={rightValue}
            header={rightHeader}
            linkField={linkField}
            direction='right'
          />
        </div>
      </div>
    </div>
  );
};

export default TableRow;
