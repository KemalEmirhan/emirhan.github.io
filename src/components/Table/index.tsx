import React from 'react';
import TableRow from './TableRow';

interface TableData {
  [key: string]: string;
}

interface TableProps {
  headers: [string, string];
  data: TableData[];
  className?: string;
  linkField?: string;
}

const Table = ({
  headers,
  data,
  className = '',
  linkField,
}: TableProps): React.ReactElement => {
  const [leftHeader, rightHeader] = headers;

  return (
    <div className={`text-black bg-white text-base ${className}`}>
      <div className='w-full max-w-4xl'>
        <div className='flex items-center pb-1 mb-1'>
          <div className='text-left text-lg font-semibold w-1/2'>
            {leftHeader}
          </div>
          <div className='text-right text-lg font-semibold w-1/2'>
            {rightHeader}
          </div>
        </div>
        <div className='border-b border-black mb-2'></div>
        <div className='space-y-1'>
          {data.map((row, index) => (
            <TableRow
              key={index}
              row={row}
              leftHeader={leftHeader}
              rightHeader={rightHeader}
              linkField={linkField}
            />
          ))}
        </div>
      </div>
    </div>
  );
};

export default Table;
