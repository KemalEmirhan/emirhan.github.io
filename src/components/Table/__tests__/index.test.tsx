import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import Table from '../';

const mockData = [
  { platform: 'GITHUB', accountName: 'httml' },
  {
    platform: 'LINKEDIN',
    accountName: 'https://www.linkedin.com/in/emirhan-kemal-kosem-259569b3/',
  },
  { platform: 'MEDIUM', accountName: 'https://medium.com/emirhankemalkosem' },
  { platform: 'TWITTER', accountName: 'https://x.com/KsemEmir' },
  {
    platform: 'INSTAGRAM',
    accountName: 'https://www.instagram.com/emirhankemalkosem/',
  },
  {
    platform: 'BLUESKY',
    accountName: 'https://bsky.app/profile/emirkosem.bsky.social',
  },
];

const mockHeaders: [string, string] = ['PLATFORM', 'ACCOUNT'];

describe('Table Component', () => {
  it('should render', () => {
    const { container } = render(
      <Table headers={mockHeaders} data={mockData} />
    );
    expect(container).toMatchSnapshot();
  });

  it('applies custom className when provided', () => {
    const customClass = 'custom-table-class';
    const { container } = render(
      <Table headers={mockHeaders} data={mockData} className={customClass} />
    );

    expect(container.firstChild).toHaveClass(customClass);
    expect(container).toMatchSnapshot();
  });

  it('renders with empty data array', () => {
    const { container, getByText } = render(
      <Table headers={mockHeaders} data={[]} />
    );

    expect(getByText('PLATFORM')).toBeInTheDocument();
    expect(getByText('ACCOUNT')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('renders links when linkField is provided', () => {
    const linkData = [
      { platform: 'GITHUB', accountName: 'https://github.com/KemalEmirhan' },
    ];
    const linkHeaders: [string, string] = ['PLATFORM', 'ACCOUNT'];

    const { container, getByRole } = render(
      <Table headers={linkHeaders} data={linkData} linkField='ACCOUNT' />
    );

    const link = getByRole('link', { name: 'https://github.com/KemalEmirhan' });
    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', 'https://github.com/KemalEmirhan');
    expect(link).toHaveAttribute('target', '_blank');
    expect(link).toHaveAttribute('rel', 'noopener noreferrer');
    expect(container).toMatchSnapshot();
  });

  it('renders regular text when linkField is not provided', () => {
    const { container, getByText, queryByRole } = render(
      <Table headers={mockHeaders} data={mockData} />
    );

    expect(getByText('GITHUB')).toBeInTheDocument();
    expect(getByText('LINKEDIN')).toBeInTheDocument();
    expect(
      getByText('https://www.linkedin.com/in/emirhan-kemal-kosem-259569b3/')
    ).toBeInTheDocument();
    expect(getByText('MEDIUM')).toBeInTheDocument();
    expect(
      getByText('https://medium.com/emirhankemalkosem')
    ).toBeInTheDocument();

    // Should not be links
    expect(queryByRole('link')).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
