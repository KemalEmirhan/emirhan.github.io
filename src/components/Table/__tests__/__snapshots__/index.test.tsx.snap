// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table Component > applies custom className when provided 1`] = `
<div>
  <div
    class="text-black bg-white text-base custom-table-class"
  >
    <div
      class="w-full max-w-4xl"
    >
      <div
        class="flex items-center pb-1 mb-1"
      >
        <div
          class="text-left text-lg font-semibold w-1/2"
        >
          PLATFORM
        </div>
        <div
          class="text-right text-lg font-semibold w-1/2"
        >
          ACCOUNT
        </div>
      </div>
      <div
        class="border-b border-black mb-2"
      />
      <div
        class="space-y-1"
      >
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              GITHUB
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              httml
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              LINKEDIN
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://www.linkedin.com/in/emirhan-kemal-kosem-259569b3/
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              MEDIUM
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://medium.com/emirhankemalkosem
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              TWITTER
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://x.com/KsemEmir
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              INSTAGRAM
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://www.instagram.com/emirhankemalkosem/
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              BLUESKY
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://bsky.app/profile/emirkosem.bsky.social
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table Component > renders links when linkField is provided 1`] = `
<div>
  <div
    class="text-black bg-white text-base "
  >
    <div
      class="w-full max-w-4xl"
    >
      <div
        class="flex items-center pb-1 mb-1"
      >
        <div
          class="text-left text-lg font-semibold w-1/2"
        >
          PLATFORM
        </div>
        <div
          class="text-right text-lg font-semibold w-1/2"
        >
          ACCOUNT
        </div>
      </div>
      <div
        class="border-b border-black mb-2"
      />
      <div
        class="space-y-1"
      >
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              GITHUB
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <a
              class="text-gray-500 hover:text-gray-800 underline flex items-center text-base"
              href="https://github.com/KemalEmirhan"
              rel="noopener noreferrer"
              target="_blank"
            >
              https://github.com/KemalEmirhan
              <svg
                aria-hidden="true"
                class="lucide lucide-arrow-up-right w-4 h-4"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 7h10v10"
                />
                <path
                  d="M7 17 17 7"
                />
              </svg>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table Component > renders regular text when linkField is not provided 1`] = `
<div>
  <div
    class="text-black bg-white text-base "
  >
    <div
      class="w-full max-w-4xl"
    >
      <div
        class="flex items-center pb-1 mb-1"
      >
        <div
          class="text-left text-lg font-semibold w-1/2"
        >
          PLATFORM
        </div>
        <div
          class="text-right text-lg font-semibold w-1/2"
        >
          ACCOUNT
        </div>
      </div>
      <div
        class="border-b border-black mb-2"
      />
      <div
        class="space-y-1"
      >
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              GITHUB
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              httml
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              LINKEDIN
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://www.linkedin.com/in/emirhan-kemal-kosem-259569b3/
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              MEDIUM
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://medium.com/emirhankemalkosem
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              TWITTER
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://x.com/KsemEmir
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              INSTAGRAM
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://www.instagram.com/emirhankemalkosem/
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              BLUESKY
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://bsky.app/profile/emirkosem.bsky.social
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table Component > renders with empty data array 1`] = `
<div>
  <div
    class="text-black bg-white text-base "
  >
    <div
      class="w-full max-w-4xl"
    >
      <div
        class="flex items-center pb-1 mb-1"
      >
        <div
          class="text-left text-lg font-semibold w-1/2"
        >
          PLATFORM
        </div>
        <div
          class="text-right text-lg font-semibold w-1/2"
        >
          ACCOUNT
        </div>
      </div>
      <div
        class="border-b border-black mb-2"
      />
      <div
        class="space-y-1"
      />
    </div>
  </div>
</div>
`;

exports[`Table Component > should render 1`] = `
<div>
  <div
    class="text-black bg-white text-base "
  >
    <div
      class="w-full max-w-4xl"
    >
      <div
        class="flex items-center pb-1 mb-1"
      >
        <div
          class="text-left text-lg font-semibold w-1/2"
        >
          PLATFORM
        </div>
        <div
          class="text-right text-lg font-semibold w-1/2"
        >
          ACCOUNT
        </div>
      </div>
      <div
        class="border-b border-black mb-2"
      />
      <div
        class="space-y-1"
      >
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              GITHUB
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              httml
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              LINKEDIN
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://www.linkedin.com/in/emirhan-kemal-kosem-259569b3/
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              MEDIUM
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://medium.com/emirhankemalkosem
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              TWITTER
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://x.com/KsemEmir
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              INSTAGRAM
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://www.instagram.com/emirhankemalkosem/
            </span>
          </div>
        </div>
        <div
          class="flex items-baseline"
        >
          <div
            class="text-left pr-2"
          >
            <span
              class="text-base uppercase"
            >
              BLUESKY
            </span>
          </div>
          <div
            class="flex-1 relative overflow-hidden h-[2px]"
          >
            <div
              class="absolute inset-0 flex items-center dot-leaders"
            />
          </div>
          <div
            class="text-right pl-2"
          >
            <span
              class="text-base uppercase"
            >
              https://bsky.app/profile/emirkosem.bsky.social
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
